File: src/core/render.ts
/**
 * Hosts file rendering
 * 
 * Converts Entry objects to /etc/hosts format lines.
 */

import type { Entry, Group, HostsFile } from "../domain/types";

/**
 * Render an Entry to /etc/hosts format.
 * 
 * Format: '<ip> <hostname> <alias1> <alias2> ...'
 * Optional comment at end if entry.comment exists: '# <comment>'
 * 
 * @param entry - The entry to render
 * @returns A single line in /etc/hosts format
 * 
 * @example
 * renderEntry({
 *   id: "01ARZ3NDEKTSV4RRFFQ69G5FAV",
 *   ip: "192.168.1.100",
 *   hostname: "devserver.local",
 *   aliases: ["devserver"],
 *   enabled: true,
 *   comment: "Development server"
 * })
 * // Returns: "192.168.1.100 devserver.local devserver # Development server"
 */
export function renderEntry(entry: Entry): string {
  const parts: string[] = [entry.ip, entry.hostname];

  // Add aliases if present
  if (entry.aliases.length > 0) {
    parts.push(...entry.aliases);
  }

  // Join with single space
  let line = parts.join(" ");

  // Add comment if present
  if (entry.comment) {
    line += ` # ${entry.comment}`;
  }

  // Disabled entries are emitted as `#`-commented lines so they remain
  // visible (and easily re-enabled by toggling the comment) in /etc/hosts.
  // See design.md:108 + rendered example :140-147. (hosts-cli-379.71)
  if (entry.enabled === false) {
    line = `# ${line}`;
  }

  return line;
}

/**
 * Wrap content in managed block markers.
 * 
 * Adds BEGIN HOSTIE and END HOSTIE markers with blank lines for readability.
 * 
 * @param content - The content to wrap (can be empty or multi-line)
 * @returns Content wrapped with markers
 * 
 * @example
 * wrapManagedBlock("192.168.1.1 example.com")
 * // Returns:
 * // # BEGIN HOSTIE
 * //
 * // 192.168.1.1 example.com
 * //
 * // # END HOSTIE
 */
export function wrapManagedBlock(content: string): string {
  return `# BEGIN HOSTIE\n\n${content}\n\n# END HOSTIE`;
}

/**
 * Recursively collect all enabled entries from a group and its subgroups.
 * 
 * @param group - The group to traverse
 * @returns Array of enabled entries (flattened)
 */
function collectAllEntries(group: Group): Entry[] {
  const entries: Entry[] = [];

  // hosts-cli-379.71: include disabled entries. renderEntry now emits them
  // as `#`-commented lines per design.md:108.
  for (const entry of group.entries) {
    entries.push(entry);
  }

  // Recursively collect from subgroups
  for (const subgroup of group.groups) {
    entries.push(...collectAllEntries(subgroup));
  }

  return entries;
}

/**
 * Render a complete HostsFile to /etc/hosts format with managed block markers.
 * 
 * Recursively traverses all groups, collects enabled entries, renders each using
 * renderEntry(), and wraps the result in managed block markers.
 * 
 * Groups are organizational only (per D4) - all enabled entries are flattened
 * into a single list regardless of group hierarchy.
 * 
 * @param hostsFile - The HostsFile to render
 * @returns Complete /etc/hosts content with BEGIN/END markers
 * 
 * @example
 * renderHostsFile({
 *   version: 1,
 *   groups: [
 *     {
 *       name: "dev",
 *       entries: [{ id: "...", ip: "192.168.1.1", hostname: "dev.local", aliases: [], enabled: true }],
 *       groups: []
 *     }
 *   ]
 * })
 * // Returns:
 * // # BEGIN HOSTIE
 * //
 * // 192.168.1.1 dev.local
 * //
 * // # END HOSTIE
 */
export function renderHostsFile(hostsFile: HostsFile): string {
  const allEntries: Entry[] = [];
  
  // Collect ALL entries (enabled and disabled) from all groups (flattened).
  // Disabled entries are rendered as `#`-commented lines by renderEntry.
  for (const group of hostsFile.groups) {
    allEntries.push(...collectAllEntries(group));
  }
  
  // Render each entry
  const lines = allEntries.map(entry => renderEntry(entry));
  
  // Join with newlines and wrap in managed block
  const content = lines.join("\n");
  return wrapManagedBlock(content);
}

File: go/internal/core/render/render.go
// Package render produces the on-disk text shape of Hostie's managed block
// inside /etc/hosts. This is the ONE canonical renderer for the Go tree — see
// "One Renderer, One Parser — Share or Pin" in docs/learnings/critical-patterns.
//
// Field separator: a SINGLE space between ip, hostname, aliases, and the
// optional " # <comment>" suffix. This matches v1 src/core/render.ts
// (renderEntry uses parts.join(" ")). /etc/hosts is whitespace-tolerant, so
// space-vs-tab is observationally irrelevant; we pin to space for byte-exact
// parity with the v1 renderer that has shipped to users.
//
// Managed-block shape: BEGIN marker, then for each group an optional
// "# group: <name>" header followed by its entries (enabled rendered raw,
// disabled rendered with a leading "# "), then END marker. There is NO
// blank-line padding inside the markers. This collapses v1's two-shape
// divergence (src/core/render.ts:wrapManagedBlock added blank lines;
// src/core/apply.ts:renderManagedBlock did not) into one canonical shape.
// The v1→v2 padding difference is documented in
// docs/go-migration/release-notes-2.0.0.md.
package render

import (
	"strings"

	"github.com/hungthai1401/hostie/go/internal/domain"
)

// Marker constants for the Hostie-managed region of /etc/hosts.
//
// These strings are matched literally (after trimming surrounding whitespace)
// by the parser in core/etchosts. They must remain byte-identical to v1's
// BEGIN_MARKER / END_MARKER in src/core/apply.ts and src/core/etchosts.ts.
const (
	BeginMarker = "# BEGIN HOSTIE"
	EndMarker   = "# END HOSTIE"
)

// RenderEntry returns a single /etc/hosts line for the given entry.
//
// Enabled:  "<ip> <hostname>[ <alias>...][ # <comment>]"
// Disabled: "# <ip> <hostname>[ <alias>...][ # <comment>]"
//
// Disabled entries are emitted as "#"-commented lines so they remain visible
// (and easily re-enabled by toggling the leading "# ") in /etc/hosts. This
// matches v1 behavior per design.md:108 (hosts-cli-379.71).
func RenderEntry(e domain.Entry) string {
	parts := make([]string, 0, 2+len(e.Aliases))
	parts = append(parts, e.IP, e.Hostname)
	parts = append(parts, e.Aliases...)

	line := strings.Join(parts, " ")
	if e.Comment != "" {
		line += " # " + e.Comment
	}
	if !e.Enabled {
		line = "# " + line
	}
	return line
}

// RenderManagedBlock returns the full managed-block string, including the
// BEGIN and END marker lines, with a trailing newline after the END marker.
//
// Shape (no blank-line padding above BEGIN or below END):
//
//	# BEGIN HOSTIE
//	# group: <name>            (only if the group has at least one entry)
//	<entry-line>
//	...
//	# END HOSTIE
//
// Each line is '\n'-terminated. Empty groups (zero entries) contribute zero
// lines — no header is emitted for them. Both enabled and disabled entries
// are rendered (disabled are "#"-commented by RenderEntry).
func RenderManagedBlock(hf *domain.HostsFile) string {
	var b strings.Builder
	b.WriteString(BeginMarker)
	b.WriteByte('\n')

	if hf != nil {
		for _, g := range hf.Groups {
			if len(g.Entries) == 0 {
				continue
			}
			b.WriteString("# group: ")
			b.WriteString(g.Name)
			b.WriteByte('\n')
			for _, e := range g.Entries {
				b.WriteString(RenderEntry(e))
				b.WriteByte('\n')
			}
		}
	}

	b.WriteString(EndMarker)
	b.WriteByte('\n')
	return b.String()
}

// RenderHostsFile is a thin convenience wrapper around RenderManagedBlock for
// callers (e.g. the golden harness, apply --dry-run) that hold a HostsFile
// value rather than a pointer. The bytes returned are byte-identical to
// RenderManagedBlock(&hf).
func RenderHostsFile(hf domain.HostsFile) string {
	return RenderManagedBlock(&hf)
}
